### PR TITLE
First attempt to use indirect access for varlen query layout

### PR DESCRIPTION
**Pull request by bringlein**
## Description

vLLM uses the so called `varlen` layout for batching of multiple requests. This format requires no padding, is more memory-efficient and hence the default (without alternative) in upstream vLLM for GPUs. 

In effect it is another indirect access, just like paging of KV Cache: Although the `query` is contiguous in memory, the beginning and end of each single request is determined by the `cu_seqlens` tensor. Therefore, to know which part of the query tensor is the right (partial) request for a given sequence id is an indirect access. 

Consequently, there was the idea to use the same trick/workaround like #205 for this side of paged attention to. 

## Related Issues

Implements #280, follow-up from #205 

## Test Plan

Unit tests and E2E offline inference. 

## Checklist

- [x] I have read the [contributing guidelines](https://torch-spyre.github.io/spyre-inference/contributing/)
- [x] My code follows the project's code style (run `bash format.sh`)
- [ ] I have added tests for my changes (if applicable)
- [ ] I have updated the documentation (if applicable)
- [x] My commits include a `Signed-off-by:` line (DCO compliance)

### CODE REVIEW

**Comment by bringlein** [comment]
### Current state

All unit tests and end-to-end inference example pass (locally). However, the list-based access is only happening for decode-only batches with size 1. 
For the other combinations (prefill, or batch size >1) there is silent data corruption happening on Spyre: Either due to `torch.cat`, `torch.stack`, or `torch.unbind`. Which I guess is related to the issue of slicing not supported on Spyre yet. 

CC @bohnstingl @jvlunteren @tdoublep

File: spyre_inference/v1/attention/backends/spyre_attn.py
Comment (bohnstingl) [inline spyre_inference/v1/attention/backends/spyre_attn.py:125]: Could we also rework this to resolve the warnings that currently appear when running this code? 
For example, by using 
```
sliced_t = output
    for dim, offset in zip(dims, offsets):
        sliced_t = torch.narrow(sliced_t, dim, offset, 1)
    sliced_t.copy_(input)
```

We have to decide whether the `copy_` is the best approach though.

File: spyre_inference/v1/attention/backends/spyre_attn.py [hunk 1]
```diff
@@ -867,7 +902,8 @@ def _reshape_and_cache(
     @_record_function("spyre_attn::online_softmax")
     def _online_softmax_attention(
         self,
-        query_cpu: torch.Tensor,
+        query_dev: torch.Tensor | None,
+        query_cpu: torch.Tensor | None,
```
Comment (bohnstingl) [inline spyre_inference/v1/attention/backends/spyre_attn.py:906]: Given the new indirect access capabilities, do we need to run these operations on CPU?

Comment (bringlein) [reply spyre_inference/v1/attention/backends/spyre_attn.py:906]: see #399

File: spyre_inference/v1/attention/backends/spyre_attn.py
Comment (bringlein) [reply spyre_inference/v1/attention/backends/spyre_attn.py:125]: I guess related to #398 and #400
